cylc/uiserver/app.py

```diff
@@ -0,0 +1,364 @@
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from pathlib import Path, PurePath
+import sys
+from typing import List
+
+from pkg_resources import parse_version
+from tornado import ioloop
+from tornado.web import RedirectHandler
+from traitlets import (
+    Float,
+    TraitError,
+    TraitType,
+    Undefined,
+    Unicode,
+    default,
+    validate,
+)
+
+from jupyter_server.base.handlers import JupyterHandler
+from jupyter_server.extension.application import ExtensionApp
+# from jupyterhub.utils import url_path_join
+import tornado
+
+from cylc.flow.network.graphql import (
+    CylcGraphQLBackend, IgnoreFieldMiddleware
+)
+
+from cylc.uiserver import (
+    __file__ as uis_pkg
+)
+from cylc.uiserver.data_store_mgr import DataStoreMgr
+from cylc.uiserver.handlers import (
+    SubscriptionHandler,
+    UIServerGraphQLHandler,
+    UserProfileHandler,
+    CylcStaticHandler
+)
+from cylc.uiserver.resolvers import Resolvers
+from cylc.uiserver.schema import schema
+from cylc.uiserver.websockets.tornado import TornadoSubscriptionServer
+from cylc.uiserver.workflows_mgr import WorkflowsManager
+
+
+class MyExtensionHandler(JupyterHandler):
+
+    @tornado.web.authenticated
+    def get(self):
+        self.write('hello cylc')
+
+    @tornado.web.authenticated
+    def post(self):
+        ...
+
+
+class PathType(TraitType):
+    """A pathlib traitlet type which allows string and undefined values."""
+
+    @property
+    def info_text(self):
+        return 'a pathlib.PurePath object'
+
+    def validate(self, obj, value):
+        if isinstance(value, str):
+            return Path(value).expanduser()
+        if isinstance(value, PurePath):
+            return value
+        if value == Undefined:
+            return value
+        self.error(obj, value)
+
+
+class CylcUIServer(ExtensionApp):
+
+    name = 'cylc'
+    app_name = 'cylc-gui'
+    default_url = "/cylc"
+    load_other_extensions = True
+    # file_url_prefix = "/render"
+    description = '''
+    Cylc - A user interface for monitoring and controlling Cylc workflows.
+    '''
+    examples = '''
+        cylc gui    # start the cylc GUI
+    '''
+    config_file_paths = list(
+        map(
+            str,
+            [
+                # base configuration - always used
+                Path(uis_pkg).parent,
+                # site configuration
+                Path('/etc/cylc/hub'),
+                # user configuration
+                Path('~/.cylc/hub').expanduser()
+            ]
+        )
+    )
```

Comment:
This registers the Cylc configuration files with Jupyter Server, they get loaded automatically.

Reply:
Does this need to support `CYLC_SITE_CONF_PATH`?